storage/repair.go

```diff
@@ -0,0 +1,278 @@
+	"github.com/m3db/m3x/log"
+
+	"github.com/uber-go/tally"
+)
+
+var (
+	errNoRepairOptions = errors.New("no repair options")
+)
+
+type recordFn func(namespace string, shard databaseShard, diffRes repair.MetadataComparisonResult)
+
+type shardRepairer struct {
+	opts    Options
+	rtOpts  retention.Options
+	session client.AdminSession
+
+	recordFn  recordFn
+	logger    xlog.Logger
+	scope     tally.Scope
+	nowFn     clock.NowFn
+	blockSize time.Duration
+	origin    topology.Host
+	replicas  int
+}
+
+func newShardRepairer(opts Options, rpOpts repair.Options) (databaseShardRepairer, error) {
+	newSessionFn := rpOpts.NewAdminSessionFn()
+	session, err := newSessionFn()
+	if err != nil {
+		return nil, err
+	}
+
+	iOpts := opts.InstrumentOptions()
+	rtOpts := opts.RetentionOptions()
+
+	r := shardRepairer{
+		opts:    opts,
+		rtOpts:  rtOpts,
+		session: session,
+
+		logger:    iOpts.Logger(),
+		scope:     iOpts.MetricsScope().SubScope("database"),
+		nowFn:     opts.ClockOptions().NowFn(),
+		blockSize: rtOpts.BlockSize(),
+		origin:    session.Origin(),
+		replicas:  session.Replicas(),
+	}
+	r.recordFn = r.recordDifferences
+
+	return r, nil
+}
+
+func (r shardRepairer) Repair(namespace string, shard databaseShard) error {
+	ctx := r.opts.ContextPool().Get()
+	defer ctx.Close()
+
+	var (
+		now   = r.nowFn()
+		start = now.Add(-r.rtOpts.RetentionPeriod()).Truncate(r.blockSize)
+		end   = now.Add(-r.rtOpts.BufferPast()).Add(-r.blockSize).Truncate(r.blockSize)
+	)
+
+	metadata := repair.NewReplicaMetadataComparer(r.replicas)
+
+	// Add local metadata
+	localMetadata, _ := shard.FetchBlocksMetadata(ctx, math.MaxInt64, 0, true, true)
+	localIter := block.NewFilteredBlocksMetadataIter(start, end, r.blockSize, localMetadata)
+	metadata.AddLocalMetadata(r.origin, localIter)
+
+	// Add peer metadata
+	peerIter, err := r.session.FetchBlocksMetadataFromPeers(namespace, shard.ID(), start, end, r.blockSize)
+	if err != nil {
+		return err
+	}
+	if err := metadata.AddPeerMetadata(peerIter); err != nil {
+		return err
+	}
+
+	// NB(xichen): only record the differences for now
+	r.recordFn(namespace, shard, metadata.Compare())
+
+	return nil
+}
+
+// TODO(xichen): log the actual differences once we have an idea of the magnitude of discrepancies
+func (r shardRepairer) recordDifferences(
+	namespace string,
+	shard databaseShard,
+	diffRes repair.MetadataComparisonResult,
+) {
+	var (
+		shardScope = r.scope.Tagged(map[string]string{
+			"namespace": namespace,
+			"shard":     strconv.Itoa(int(shard.ID())),
+		})
+		totalScope        = shardScope.Tagged(map[string]string{"resultType": "total"})
+		sizeDiffScope     = shardScope.Tagged(map[string]string{"resultType": "sizeDiff"})
+		checksumDiffScope = shardScope.Tagged(map[string]string{"resultType": "checksumDiff"})
+	)
+
+	// Record total number of series and total number of blocks
+	totalScope.Counter("repair.series").Inc(diffRes.NumSeries)
+	totalScope.Counter("repair.blocks").Inc(diffRes.NumBlocks)
```

Comment:
Unfortunately these depend on the `shardScope`, which are namespace+shard specific. We could create all of them upfront and store them in a map keyed off of namespace, shard combination but seems overkill if we only run repair once every 2 hours